lib/df-detection.c

```diff
@@ -0,0 +1,209 @@
+#include <linux/df-detection.h>
+#include <linux/debugfs.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/random.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+#include <linux/sys.h>
+#include <linux/types.h>
+
+void dfetch_add_address(const void *addr, size_t len, unsigned long caller,
+		 void *kernel_addr)
+{
+	if (!in_task() || current->dfetch_addresses == NULL || current->dfetch_pairs == NULL ||
+	    addr > TASK_SIZE)
+		return;
+	if (len > MAX_LEN || addr == 0)
+		return;
+	if (current->num_read >= current->ranges_size && current->ranges_size < DFETCH_MAX_RECORDS) {
+		struct dfetch_address_range *temp =
+		    (struct dfetch_address_range *)krealloc(
```

Comment:
Nit: I don't think using spaces for indentation is justified here.
Try running clang-format on this file.